### ISSUE TITLE
aws_default_vpc_dhcp_options should not contain any specific information

### ISSUE THREAD

**Issue by isen-ng** [traveloka/terraform-aws-vpc#23]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* ***If you are interested in working on this issue or have submitted a pull request, please leave a comment!***

***

### Summary

According to [default_vpc_dhcp_options](https://www.terraform.io/docs/providers/aws/r/default_vpc_dhcp_options.html), each region can only have 1 `default_vpc_dhcp_options`, and terraform does not create but instead adopts. This means that, if this resource is ever invoked in different stacks, the same resource will be updated.

This may create confusion as to why this resource is updated when nothing has changed in 1 particular stack.

This resource should not contain any information specific to a stack. In fact, I think this resource doesn't even need to be created, according to the documentation:

> Terraform does not create this resource, but instead "adopts" it into management.

### Module Version

0.3.0

### Terraform Version and Terraform AWS Provider Version

```
Terraform v0.11.14
+ provider.aws v2.43.0
```

### Affected Resource(s)

* aws_default_vpc_dhcp_options

### Terraform Configuration Files

```hcl
# Provides a resource to manage the default AWS DHCP Options Set in the current region.
resource "aws_default_vpc_dhcp_options" "this" {
  depends_on = ["aws_vpc.this"]

  tags = "${merge(
    local.common_tags,
    map("Name", format("%s-default-dopt", var.vpc_name)),
    map("Description", format("Default AWS DHCP options set for %s VPC", var.vpc_name)))}"
}
```

### Expected Behavior

There should be no changes to aws_default_vpc_dhcp_options when the stack has not changed.

### Actual Behavior

There are changes to aws_default_vpc_dhcp_options when the stack has not changed, because this resource can be updated by any other stack.

### Steps to Reproduce

1. `terraform apply`

**Comment by sshvetsov** [comment]
Hi @isen-ng, If I understand the issue correctly, our VPC module assumes that there is only one VPC in a given region and tries to rely on `aws_default_vpc_dhcp_options` resource, instead of creating a dedicated `aws_vpc_dhcp_options` resource for each VPC. If you're creating multiple VPC in one region, this causes problems, because the same `aws_default_vpc_dhcp_options` resource is being assigned to and updated by different VPCs. Is my understanding correct?

**Comment by isen-ng** [comment]
> our VPC module assumes that there is only one VPC in a given region and tries to rely on aws_default_vpc_dhcp_options resource, instead of creating a dedicated aws_vpc_dhcp_options resource for each VPC

Yes, however, I don't know if there's anything actually being relied on because there is nothing being customized in `aws_default_vpc_dhcp_options` except metadata like Name and Description.

>  If you're creating multiple VPC in one region, this causes problems, because the same aws_default_vpc_dhcp_options resource is being assigned to and updated by different VPCs. 

That's the basic idea. If this resource is set by another stack, whether it's our VPC module, or any other stack that modifies this resource, we will run into this cosmetic (as there are no customizations right now) issue.

**Comment by sshvetsov** [comment]
Just to be clear, the DHCP options are not expected to be modified outside of the VPC module. If you have such requirements, please do explain your use case.

I see two possible approaches:
1. VPC module is responsible for managing the DHCP options of the VPC.
2. VPC module is not responsible for managing the DHCP options of the VPC.

Are you proposing to completely remove DHCP options handling from the VPC module?

**Comment by isen-ng** [comment]
> Just to be clear, the DHCP options are not expected to be modified outside of the VPC module. If you have such requirements, please do explain your use case.

That may be true from a policy perspective, but the nature of this resource allows it to be modified, whether intentionally, or unintentionally, using terraform, from a different stack. 

> I see two possible approaches:
> 
>   1. VPC module is responsible for managing the DHCP options of the VPC.
>   2. VPC module is not responsible for managing the DHCP options of the VPC.

I propose to remove the management of the default DHCP options `default_vpc_dhcp_options`, of the region from within the VPC. If there is a requirement, create a new resource `aws_vpc_dhcp_options` to manage the dhcp options of the VPC.

`default_vpc_dhcp_options`: Default DHCP options, there is only 1 per region
`aws_vpc_dhcp_options`: Per VPC DHCP options. There can be many per region.

**Comment by sshvetsov** [comment]
It is not really a policy, just part of our original assumptions that each AWS account in our Org will have only one VPC, as we create one account per Product Domain per Environment. All resources are expected to be modified, but **very** few resources are expected to be modified outside of the stacks that originally created them. That would be a bad practice and I do not recommend it for most use cases.

I agree `default_vpc_dhcp_options` doesn't belong in the VPC module if multiple VPCs are created in the single region. So far only your team seems to have such requirement in our Org, but it makes sense to drop this resource from the module, considering we can do it with minimal impact to the existing VPCs.

Since `aws_vpc_dhcp_options` will typically be used to define the domain name used in the VPC and may not always be created at the same time as VPC, it might be a good idea to add it and `aws_vpc_dhcp_options_association` as a sub-module in this module, to ensure the resource is tagged per our policies and is not left orphaned after VPC removal.

I think you request has a merit. We'll discuss on the best way to handle it.
